### PR TITLE
add Half support for logsigmoid, threshold, elu, gelu, hardtanh, hardsigmoid, hardswish, hardshrink, softshrink, leakyrelu, softplus, glu, silu, mish, and prelu on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98745

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -2108,12 +2108,13 @@ def test_threshold_int(self):
         expected = torch.tensor([99, 99, 99, 99, 1, 2, 3])
         self.assertEqual(F.threshold(x, 0, 99), expected)
 
-    def test_threshold_bfloat16(self):
+    def test_threshold_bfloat16_half(self):
```
Comment (jgong5) [inline test/test_nn.py:2111]: Can we use `@dtype` decorator?

File: test/test_nn.py [hunk 2]
```diff
@@ -11130,12 +11131,12 @@ def test_multi_margin_loss_errors(self, device):
                                                                   torch.zeros(3, device=device)))
 
     @onlyCPU
-    def test_activations_bfloat16_cpu(self, device):
-        def test_bfloat16(fn, device, inp_dims, prec):
-            # bfloat16 compute
-            input = torch.randn(inp_dims, dtype=torch.bfloat16, device=device, requires_grad=True)
+    def test_activations_bfloat16_half_cpu(self, device):
```
Comment (jgong5) [inline test/test_nn.py:11134]: ditto: dtype decorator?

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 3]
```diff
@@ -185,17 +185,22 @@ void elu_kernel(TensorIteratorBase& it, const Scalar& alpha, const Scalar& scale
...
-            auto res0 = Vectorized<float>::blendv(((a0 * negiptcoef_vec).exp() - one_vec) * negcoef_vec, a0 * poscoef_vec, cmp0);
-            auto res1 = Vectorized<float>::blendv(((a1 * negiptcoef_vec).exp() - one_vec) * negcoef_vec, a1 * poscoef_vec, cmp1);
-            return convert_from_float<scalar_t>(res0, res1);
+            res0 = Vectorized<float>::blendv(((a0 * negiptcoef_vec).exp() - one_vec) * negcoef_vec, a0 * poscoef_vec, cmp0);
           }
+          if (!cmp1.zero_mask()) {
+            res1 = a1 * poscoef_vec;
+          } else {
+            res1 = Vectorized<float>::blendv(((a1 * negiptcoef_vec).exp() - one_vec) * negcoef_vec, a1 * poscoef_vec, cmp1);
+          }
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/Activation.cpp]: Would it be simpler to use a lambda function to implement it? e.g.
```c++
auto get_res_masked = [&](Vectorized<float> cmp, Vectorized<float> a) {
  return !cmp.zero_mask() ? a * poscoef_vec : Vectorized<float>::blendv(((a * negiptcoef_vec).exp() - one_vec) * negcoef_vec, a * poscoef_vec, cmp);
}
```

**Review by jgong5 — APPROVED** [review]
LGTM in general. Some small nits, please check.

**Review by mingfeima — APPROVED** [review]

File: aten/src/ATen/native/cpu/Activation.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/Activation.cpp]: Good idea.  I will use lambda function to simplify the code.

File: test/test_nn.py (hunk 1 above)
Comment (CaoE) [reply test/test_nn.py:2111]: Can't use @dtype here because it is in the scope of TestNN not TestNNDeviceType.

File: test/test_nn.py (hunk 2 above)
Comment (CaoE) [reply test/test_nn.py]: Use dtype decorator

**Comment by CaoE** [comment]
@ngimel Could you please help to review this PR? Thank you.

**Comment by CaoE** [comment]
@ngimel Could you please help to review this PR? Thanks.

**Comment by CaoE** [comment]
@Skylion007  Could you please help to review this PR? Thank you.

**Comment by CaoE** [comment]
@ngimel Could you please help to review this PR? Thanks.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**